From: Heiko Carstens <heiko.carstens@de.ibm.com>
Subject: s390/cpuinfo: show facilities as reported by stfle
Patch-mainline: v4.12-rc1
Git-commit: 157467ba9fb7e379f0540707dd89111de441e45e
References: bnc#1076849, LTC#163741

Add a new line to /proc/cpuinfo which shows all available facilities
as reported by the stfle instruction:

> cat /proc/cpuinfo
...
facilities      : 0 1 2 3 4 6 7 ...
...

Reviewed-by: Peter Oberparleiter <oberpar@linux.vnet.ibm.com>
Signed-off-by: Heiko Carstens <heiko.carstens@de.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/processor.c |   14 ++++++++++++++
 1 file changed, 14 insertions(+)

--- a/arch/s390/kernel/processor.c
+++ b/arch/s390/kernel/processor.c
@@ -14,6 +14,7 @@
 #include <linux/seq_file.h>
 #include <linux/delay.h>
 #include <linux/cpu.h>
+#include <linux/bitops.h>
 #include <asm/elf.h>
 #include <asm/lowcore.h>
 #include <asm/param.h>
@@ -36,6 +37,18 @@ void __cpuinit cpu_init(void)
 	memset(idle, 0, sizeof(*idle));
 }
 
+static void show_facilities(struct seq_file *m)
+{
+	unsigned int bit;
+	long *facilities;
+
+	facilities = (long *)&S390_lowcore.stfle_fac_list;
+	seq_puts(m, "facilities      :");
+	for_each_set_bit_left(bit, facilities, MAX_FACILITY_BIT)
+		seq_printf(m, " %d", bit);
+	seq_putc(m, '\n');
+}
+
 /*
  * show_cpuinfo - Get information on one CPU for use by procfs.
  */
@@ -60,6 +73,7 @@ static int show_cpuinfo(struct seq_file
 			if (hwcap_str[i] && (elf_hwcap & (1UL << i)))
 				seq_printf(m, "%s ", hwcap_str[i]);
 		seq_puts(m, "\n");
+		show_facilities(m);
 	}
 	get_online_cpus();
 	if (cpu_online(n)) {
